### PR TITLE
perf(code): stream SARIF conversion [IDE-1940]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

Streams SARIF result decoding in the Snyk Code conversion path so large Code scan payloads do not require full result-array materialization before issue conversion.

This is split PR 2 from `performance/IDE-1940_optimize-memory-footprint`. It is temporarily stacked on PR #1250 so the pre-push/test-hook stabilizers from the fixture harness PR are present; after PR #1250 lands this PR can be retargeted to `main`.

### Split Context

```mermaid
flowchart LR
  main[origin/main] --> pr1["PR #1250: test fixture + megaproject harness"]
  pr1 --> pr2["PR #1251: Code SARIF streaming"]
  main --> pr3["PR TBD: OSS CLI streaming"]
  main --> pr4["PR TBD: OSS conversion caches"]
  main --> pr5["PR TBD: IssueCache index + memory backend"]
  pr5 --> pr6["PR TBD: Bolt backend + scanner integration"]
  pr6 --> pr7["PR TBD: Workspace/product cache clearing"]
  main --> pr8["PR TBD: Small hotpath optimizations"]
  main --> pr9["PR TBD: Perf docs + profiling history"]
```

### Scope

- Adds `conversion_stream.go` to token-walk `runs[0].results` with `json.Decoder`.
- Keeps streaming and full-unmarshal conversion behavior aligned through shared result append logic.
- Adds conversion tests for parity, empty runs, and malformed result handling.

### Verification

- `go test -count=1 -v ./infrastructure/code` (`build/pr2_focused_code_tests.log`)
- `make lint` (`build/pr2_make_lint.log`)
- Pre-push hook: `lint-fix` and unit tests passed during `git push --set-upstream origin perf/IDE-1940-code-sarif-streaming`

### Checklist

- [x] Tests added and all succeed
- [ ] Regenerated mocks, etc. (`make generate`)
- [x] Linted (`make lint-fix`)
- [ ] README.md updated, if user-facing
- [x] License file updated, if new 3rd-party dependency is introduced

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduce streaming SARIF result decoding to reduce memory footprint.

- Implement new tests for streaming conversion parity and edge cases.

- Refactor conversion logic to support per-result processing.

- Fallback to full JSON unmarshalling when duplicate SARIF keys are encountered.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A[Full JSON Unmarshal] --> B{Process Results};
  B --> C(Convert to Issues);
  A --> D{Stream SARIF Decoding};
  D --> E{Process Results (one by one)};
  E --> F(Convert to Issues);
  G[New Test Cases] --> B;
  G --> D;
  H{Duplicate Keys Detected} --> A;
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>conversion.go</strong><dd><code>Refactor SARIF conversion for streaming and error handling</code></dd></summary>
<hr>

infrastructure/code/conversion.go

<ul><li>Refactor <code>ConvertSARIFJSONToIssues</code> to use streaming JSON decoding.<br> <li> Introduce a new method <code>appendIssuesForResult</code> for processing individual <br>SARIF results.<br> <li> Implement fallback logic to full JSON unmarshalling when duplicate <br>SARIF keys are encountered, to preserve encoding/json semantics.<br> <li> Enhance error handling by joining multiple conversion errors.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1251/changes#diff-cf02ee8aead14021f3336c6c9c4fe368ce974be4c41bc08bdaa2364326aede89">+38/-9</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>conversion_stream.go</strong><dd><code>Implement SARIF JSON streaming decoder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

infrastructure/code/conversion_stream.go

<ul><li>Add new file containing the core streaming logic for SARIF JSON.<br> <li> Implement <code>json.Decoder</code> based token-walking for efficient result <br>processing.<br> <li> Define <code>sarifRunHead</code> and <code>sarifDocumentHead</code> structs for partial parsing.<br> <li> Handle duplicate SARIF keys by returning <br><code>errDuplicateSARIFStreamingKey</code>.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1251/changes#diff-02358a5e5905ef12f9be2122af2217e5ce895c1f07494ee3d40569a2922c8ef7">+177/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>convert.go</strong><dd><code>Modularize SARIF conversion logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

infrastructure/code/convert.go

<ul><li>Refactor <code>toIssues</code> method to delegate result processing to a new <br><code>appendIssuesForResult</code> method.<br> <li> This change modularizes the conversion logic, supporting the new <br>streaming approach.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1251/changes#diff-c487190192b10556a66dd6213400d3084215c69eb2c4f547285fe7b4ab61c0f4">+127/-113</a></td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>conversion_test.go</strong><dd><code>Add unit tests for SARIF streaming conversion</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

infrastructure/code/conversion_test.go

<ul><li>Add new test file for SARIF conversion streaming functionality.<br> <li> Include tests to verify parity between streaming and full <br>unmarshalling.<br> <li> Add tests for empty runs, malformed results, and duplicate keys <br>fallback.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1251/changes#diff-6a35334a7e6bba7a5432e8a4e24f5779c42a87272b84ac7e49e2df3dabd7118c">+279/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by bastiandoetsch** [comment]
Addressed the Code SARIF streaming review findings:

- New conversion tests now use normal unit-test setup instead of smoke-test setup, so they run in regular unit test jobs.
- Added duplicate `runs` / duplicate `results` regression coverage.
- Streaming conversion now falls back to the full `encoding/json` unmarshal path when duplicate SARIF keys would otherwise diverge from old semantics.
- Rebased onto the latest #1250 base after `origin/main` was merged there.

Verification:
- `go test -count=1 -v ./infrastructure/code`
- `make lint`
- pre-push `lint-fix` and unit tests passed during force-with-lease push

File: infrastructure/code/conversion.go [hunk 1]
```diff
@@ -32,20 +33,48 @@ import (
...
+	sarifResponse.Sarif.Runs = []codeClientSarif.Run{run}
 
 	converter := SarifConverter{sarif: sarifResponse, logger: logger, hoverVerbosity: hoverVerbosity, engine: engine}
+	ruleLink := createRuleLink()
+	baseDir := types.FilePath(basePath)
 
-	// Convert with provided base path (or empty for relative paths)
-	issues, err := converter.toIssues(types.FilePath(basePath))
+	var issues []types.Issue
+	var errs error
```
Comment (acke) [inline infrastructure/code/conversion.go]: We use err, errs, err2 in the logic below. Would it be helpful to have more descriptive names for these errors to avoid confusion?

File: infrastructure/code/conversion.go [hunk 2]
```diff
@@ -32,20 +33,48 @@ import (
 // This is a simplified version for use by MCP and other tools that need conversion without full scanner
 // basePath is the absolute path where the scan was run (optional - if empty, paths remain relative)
 func ConvertSARIFJSONToIssues(engine workflow.Engine, logger *zerolog.Logger, hoverVerbosity int, sarifJSON []byte, basePath string) ([]types.Issue, error) {
-	var sarifResponse codeClientSarif.SarifResponse
-
-	err := json.Unmarshal(sarifJSON, &sarifResponse.Sarif)
-	if err != nil {
+	var head sarifDocumentHead
+	if err := json.Unmarshal(sarifJSON, &head); err != nil {
```
Comment (acke) [inline infrastructure/code/conversion.go:37]: The whole document is scanned twice.

`json.Unmarshal(sarifJSON, &head)`
encoding/json must parse the entire input to finish unmarshaling. For each run you only bind tool and properties; the results array is still parsed and skipped (not turned into Result structs), but those bytes are still read and walked once.

`streamFirstRunResults`
Starts a json.Decoder at byte 0 and walks the document again: skip top-level keys until runs, enter runs, walk the first run until results, then Decode each result. Remaining runs are still consumed (finishRunsArrayAfterFirstRun). So the full payload is traversed again.

So the “double parse” is two full passes over the JSON, not “small head twice.”

What is not doubled: results[*] are unmarshaled into codeClientSarif.Result only on the streaming pass — the head Unmarshal never allocates/decodes those objects.

If duplicate keys force the fallback, there is a third full parse: full json.Unmarshal into SarifResponse.

**Review by acke — APPROVED** [review]
Two comments/questions, if the behavior is what we want, then it's ok to merge.
